### ISSUE TITLE
Training Multiple epoch 

### ISSUE THREAD

**Issue by raju249** [OpenNMT/OpenNMT#203]
Hi team,

I wanted to know that is it possible to train the model using 2 parallel epoch process ?
For example in one bash process I might have epoch 2 going on for whole 64k iterations and I would like to start a new bash process with epoch 3.

Is it possible ? or each epoch depends on model from previous epoch ?

**Comment by guillaumekln** [comment]
Hi,

I'm not sure I understand the use case. What would you do with the two trained models?

**Comment by raju249** [comment]
I wanted to speed up the process.
My server is AWS g2.2xLarge instance but takes a day to complete one epoch.
So I thought this might be the way to speed up the process.
Can you please share any other ideas if you have .?

**Comment by guillaumekln** [comment]
Neural network optimization is a sequential process. You could instead:

* use [multiple GPUs](http://opennmt.net/OpenNMT/training/multi_gpu/)
* come up with a clever mix of [data sampling](http://opennmt.net/OpenNMT/training/sampling/) and learning rate [decay strategy](http://opennmt.net/OpenNMT/training/decay/)
* reduce your model size

**Comment by raju249** [comment]
Thanks @guillaumekln.
A bit new to ML and Neural Networks.
Thanks for your help though.
I will close this issue for now 😊

**Comment by raju249** [comment]
Hi @guillaumekln 
One quick question,
For a training on a given epoch number 'x', which previous epochs does it use for sequential process ?

**Comment by shahbazsyed** [comment]
Hi @raju249 
The model is continuously updated for each epoch. You can observe the perplexity of the model changing with each epoch; this is due to the weights and parameters being updated by the errors/information learnt from the iterations so far. So if yo observe that at some epoch **n** your model has the lowest/desirably low perplexity, you might choose to stop the training and use this as your baseline model on the test set.

tl;dr the model is made up from all the previous epochs, and not just the immediate previous epoch

**Comment by raju249** [comment]
Thanks @shahbazsyed

**Comment by raju249** [comment]
Can you explain me what exactly is perplexity ? @shahbazsyed

**Comment by shahbazsyed** [comment]
@raju249 Perplexity can be intuitively seen as the difference between two probability distributions. In the case of NMT you have two distributions for two languages (e.g English -> German). As the model learns, it tries to minimize the distance/ difference between these two distributions, which ultimately allow it to give out accurate/nearly accurate translations in the test phase. 

You can see the perplexity between your source and targets being continuously logged in the training phase. The hope is it gradually decreases with every epoch. 

![image](https://cloud.githubusercontent.com/assets/4356098/25306234/477be7b6-2789-11e7-8f8d-f7254246aaf7.png)

**Comment by raju249** [comment]
Thanks a lot @shahbazsyed
